roles/ceph-nfs/tasks/pre_requisite_non_container.yml

```diff
@@ -1,120 +1,154 @@
+            state: "{{ (upgrade_ceph_packages|bool) | ternary('latest','present') }}"
+          when:
+            - nfs_file_gw
+        - name: install redhat nfs-ganesha-rgw and ceph-radosgw packages
+          package:
+            name: "{{ item }}"
+            state: "{{ (upgrade_ceph_packages|bool) | ternary('latest','present') }}"
+          with_items:
+            - nfs-ganesha-rgw
+            - ceph-radosgw
+          when:
+            - nfs_obj_gw
   when:
     - ansible_os_family == 'RedHat'
-    - nfs_ganesha_stable
-    - ceph_origin == 'repository'
-    - ceph_repository == 'community'
-
-- name: fetch nfs-ganesha red hat development repository
-  uri:
-    url: https://shaman.ceph.com/api/repos/nfs-ganesha/next/latest/{{ ansible_distribution | lower }}/{{ ansible_distribution_major_version }}/flavors/{{ nfs_ganesha_flavor }}/repo
-    return_content: yes
-  register: nfs_ganesha_dev_yum_repo
-  when:
-    - ansible_os_family == 'RedHat'
-    - nfs_ganesha_dev
-    - ceph_origin == 'repository'
-    - ceph_repository == 'dev'
 
-- name: add nfs-ganesha development repository
-  copy:
-    content: "{{ nfs_ganesha_dev_yum_repo.content }}"
-    dest: /etc/yum.repos.d/nfs-ganesha-dev.repo
-    owner: root
-    group: root
-    backup: yes
-  when:
-    - ansible_os_family == 'RedHat'
-    - nfs_ganesha_dev
-    - ceph_origin == 'repository'
-    - ceph_repository == 'dev'
-
-- name: add nfs-ganesha stable repository
-  apt_repository:
-    repo: "deb {{ nfs_ganesha_stable_deb_repo }} {{ ceph_stable_distro_source | default(ansible_lsb.codename) }} main"
-    state: present
-    update_cache: no
-  register: add_ganesha_apt_repo
-  when:
-    - ansible_os_family == 'Debian'
-    - nfs_ganesha_stable
-    - ceph_origin == 'repository'
-    - ceph_repository == 'community'
-
-- name: update apt cache
-  apt:
-    update_cache: yes
-  register: update_ganesha_apt_cache
-  retries: 5
-  delay: 2
-  until: update_ganesha_apt_cache | success
-  when:
-    - ansible_os_family == 'Debian'
-    - nfs_ganesha_stable
-    - ceph_origin == 'repository'
-    - ceph_repository == 'community'
-    - add_ganesha_apt_repo | changed
-
-- name: fetch nfs-ganesha development repository
-  uri:
-    url: https://shaman.ceph.com/api/repos/nfs-ganesha/next/latest/{{ ansible_distribution | lower }}/{{ ansible_distribution_release }}/flavors/{{ nfs_ganesha_flavor }}/repo
-    return_content: yes
-  register: nfs_ganesha_apt_repo
-  when:
-    - ansible_os_family == 'Debian'
-    - nfs_ganesha_dev
-    - ceph_origin == 'repository'
-    - ceph_repository == 'dev'
-
-- name: add nfs-ganesha development repository
-  copy:
-    content: "{{ nfs_ganesha_dev_apt_repo.content }}"
-    dest: /etc/apt/sources.list.d/nfs-ganesha-dev.list
-    owner: root
-    group: root
-    backup: yes
+- name: debian related tasks
+  block:
+    - name: debian repository handling
+      block:
+        - name: stable repos specific tasks
+          block:
```

Comment:
@leseb I think having such consecutive `block` directives can be confusing and make the code a bit difficult to read. 

What do you think ?

Reply:
Personally, I like the nesting because it brings similar tasks together. Otherwise, the repetition of conditions makes it difficult to see that the tasks are related and reading the playbooks becomes more difficult. For example -

**Style 1**
```
- task 1
- when
  - a
  - b

- task 2
- when
  - a
  - b
```

**Style 2**
```
- block
    - task1
    - task2
- when
  - a
  - b
```

Reply:
@rishabh-d-dave in that case I wonder if we better to put these tasks together in one included file.
yes, block can be nice, but I'm still thinking too many nested block can be difficult to read at some point.

Reply:
> in that case I wonder if we better to put these tasks together in one included file.

That would certainly reduce one or two levels of nesting. And, the playbooks where the these tasks would be included would certainly look much more readable. But on the downside, the playbook names might become too long. For example,  `pre_requisite_non_container.yml`  would split into `pre_requisite_non_container_for_debian_based_systems.yml` and `pre_requisite_non_container_for_red_hat_based_systems.yml`

Alternatively, we can move `when` before the block begins which might substantially reduce the confusion -

**Current style**
```
- name: block 0
  block:
    - name: task1
    - name: task2
       block:
        - name: task2.1
        - name: task2.2
        - name: task2.3
      when: cond #2
    - name: task3
    - name: task4
    - name: task5
  when cond #1
```

**Proposed style**
```
- name: block 0
  when: cond #1
  block:
    - name: task1
    - name: task2
       block:
       when: cond #2
        - name: task2.1
        - name: task2.2
        - name: task2.3
    - name: task3
    - name: task4
    - name: task5
```

Reply:
Just to clarify a bit. I'm not 100% against using block directive.
I'm just unsure about the confusion that it might bring at some point when too many level of nested block (I'm thinking when >2) especially when you have a long list of task inside each of them.

I think it's just a matter of a good compromise between too many duplicate `when` and too many nested block.

file name is not a real issue, I think. eg:
`pre_requisite_non_container_for_debian_based_systems.yml` could be `pre_requisite_non_container_[family].yml` where `[family]` would be the family name.